### ISSUE TITLE
Charmcraft 3 no longer respects legacy bases architectures

### ISSUE THREAD

**Issue by dragomirp** [canonical/charmcraft#1783]
### Bug Description

Building a charm on a non-amd64 architecture using the legacy Charmcraft 2 `bases` notation on Charmcraft 3 ignores the architecture and defaults to amd64 despite building on a different platform.

Example for this is the [postgresql-tes-app](https://github.com/canonical/postgresql-test-app) packing and CI tests passed fine, using a locally packed charm, but both amd64 and arm64 revisions got published as amd64 revisions.

The `manifest.yaml` inside the published arm64 revison (203) is:
```
charmcraft-version: 3.1.1
charmcraft-started-at: '2024-08-02T10:30:16.196847'
bases:
- name: ubuntu
  channel: '22.04'
  architectures:
  - amd64
- name: ubuntu
  channel: '22.04'
  architectures:
  - arm64
analysis:
  attributes:
  - name: language
    result: python
  - name: framework
    result: operator
```

The packed charm is correctly marked as arm64

### To Reproduce

* Pack postgresql-test-app on arm64
* Publish on charmhub
* Deploy on amd64

### Environment

OS: Ubuntu 22.04 (Github runner)
Charmcraft: 3.1.1

### charmcraft.yaml

```shell
# Copyright 2023 Canonical Ltd.
# See LICENSE file for licensing details.

type: charm
bases:
  - name: ubuntu
    channel: "22.04"
    architectures: [amd64]
  - name: ubuntu
    channel: "22.04"
    architectures: [arm64]
parts:
  charm:
    override-pull: |
      craftctl default
      if [[ ! -f requirements.txt ]]
      then
          echo 'ERROR: Use "tox run -e build" instead of calling "charmcraft pack" directly' >&2
          exit 1
      fi
    charm-strict-dependencies: true
  libpq:
    build-packages:
      - libpq-dev
    plugin: dump
    source: /usr/lib/
    source-type: local
    prime:
      - lib/
    organize:
      "*-linux-gnu/libpq.so*": lib/
```

### Relevant log output

```shell
Packing...
Packing charm postgresql-test-app_ubuntu-22.04-arm64.charm
Packed postgresql-test-app_ubuntu-22.04-arm64.charm
build-dev: commands_post[0]> mv requirements.txt requirements-last-build.txt
  build-dev: OK (304.68=setup[0.95]+cmd[0.66,303.07,0.00] seconds)
  congratulations :) (304.74 seconds)
```

**Comment by lengau** [comment]
Thanks for the report! Can you share the `metadata.yaml` file from the charm being packed on arm64? I don't have an arm64 machine to pack on, and cross-compiling it made the build from my machine look correct.

**Comment by lengau** [comment]
Root cause here: we're generating the CharmMetadata from charmcraft.yaml without replacing the bases to use the current build plan.

**Comment by lengau** [comment]
This should be fixed in 3.1.2.

**Comment by mmkay** [comment]
hey @lengau, any chances it will land on `stable` soon?
